### PR TITLE
[Enhancement]hyperscan with vectorization

### PR DESCRIPTION
**Pull request by zombee0**
Why I'm doing:
regexp_replace costs a lot cpu, we want to optimize it.

What I'm doing:
use hyperscan with vectorization

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/string_functions.cpp [hunk 1]
```diff
@@ -14,12 +14,13 @@
 
 #include "exprs/string_functions.h"
 
-#include <hs/hs.h>
+#include "column/vectorized_fwd.h"
+#include "formats/parquet/column_chunk_reader.h"
```
Comment (stdpain) [inline be/src/exprs/string_functions.cpp:18]: why include paquet chunk reader?

Comment (zombee0) [reply be/src/exprs/string_functions.cpp]: maybe introduced by IDE, fixed.

File: be/src/exprs/string_functions.cpp [hunk 2]
```diff
@@ -3030,61 +3027,6 @@ int StringFunctions::index_of(const char* source, int source_count, const char*
     return -1;
 }
 
-struct StringFunctionsState {
```
Comment (dirtysalt) [inline be/src/exprs/string_functions.cpp:3033]: Why move it to header file? Does not that will increase compilation time?

**Comment by dirtysalt** [comment]
What does this result mean?

What the meaning of first/second argument? 

```
/*
    -----------------------------------------------------------------------------------
    Benchmark                                         Time             CPU   Iterations
    -----------------------------------------------------------------------------------
    BM_HyperScan_Eval/5/1/iterations:10000      1427727 ns      1427629 ns        10000
    BM_HyperScan_Eval/5/0/iterations:10000      1082300 ns      1082400 ns        10000
    BM_HyperScan_Eval/10/1/iterations:10000     1002910 ns      1002870 ns        10000
    BM_HyperScan_Eval/10/0/iterations:10000      587874 ns       587929 ns        10000
    BM_HyperScan_Eval/20/1/iterations:10000      752652 ns       752632 ns        10000
    BM_HyperScan_Eval/20/0/iterations:10000      359769 ns       359775 ns        10000
    BM_HyperScan_Eval/50/1/iterations:10000      566782 ns       566772 ns        10000
    BM_HyperScan_Eval/50/0/iterations:10000      183832 ns       183852 ns        10000
    BM_HyperScan_Eval/100/1/iterations:10000     497867 ns       497860 ns        10000
    BM_HyperScan_Eval/100/0/iterations:10000     100563 ns       100588 ns        10000
     */
```

File: be/src/exprs/string_functions.cpp (hunk 2 above)
Comment (zombee0) [reply be/src/exprs/string_functions.cpp:3033]: we need this struct for ut and benchmark, so move it to header and included by ut and bench.

**Comment by zombee0** [comment]
> What the meaning of first/second argument?
>  the first argument means there is a character to be replaced in how many characters, like one of five, one of ten...
the second argument means use vectorization or not, 1 means not (old way), 0 means yes.
> What does this result mean?
> as the meaning of argument above, we see that with vectorization we can get 1.3-5 times faster.
> ```
> /*
>     -----------------------------------------------------------------------------------
>     Benchmark                                         Time             CPU   Iterations
>     -----------------------------------------------------------------------------------
>     BM_HyperScan_Eval/5/1/iterations:10000      1427727 ns      1427629 ns        10000
>     BM_HyperScan_Eval/5/0/iterations:10000      1082300 ns      1082400 ns        10000
>     BM_HyperScan_Eval/10/1/iterations:10000     1002910 ns      1002870 ns        10000
>     BM_HyperScan_Eval/10/0/iterations:10000      587874 ns       587929 ns        10000
>     BM_HyperScan_Eval/20/1/iterations:10000      752652 ns       752632 ns        10000
>     BM_HyperScan_Eval/20/0/iterations:10000      359769 ns       359775 ns        10000
>     BM_HyperScan_Eval/50/1/iterations:10000      566782 ns       566772 ns        10000
>     BM_HyperScan_Eval/50/0/iterations:10000      183832 ns       183852 ns        10000
>     BM_HyperScan_Eval/100/1/iterations:10000     497867 ns       497860 ns        10000
>     BM_HyperScan_Eval/100/0/iterations:10000     100563 ns       100588 ns        10000
>      */
> ```

**Review by stdpain — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Comment by zombee0** [comment]
https://github.com/Mergifyio backport branch-3.2

**Pull request merged**